File: showrecommender.cpp
#include "showrecommender.h"
#include <QRandomGenerator>

ShowRecommender::ShowRecommender(TraktClient* client, StateStore* state, QObject* parent)
    : QObject(parent),
      client_(client),
      state_(state)
{
}

void ShowRecommender::ensureLoaded()
{
    if (!candidates_.isEmpty())
        return;

    if (!client_ || !client_->isValid())
        return;

    inLists_ = client_->fetchAllListShowIds();
    watched_ = client_->fetchWatchedShowIds();

    QVector<Show> popular = client_->fetchPopularShows();
    QVector<Show> trending = client_->fetchTrendingShows();

    candidates_ = popular;
    for (const Show& s : trending)
        candidates_.append(s);

    // Uniform random order
    std::shuffle(candidates_.begin(), candidates_.end(),
                 *QRandomGenerator::global());

    index_ = 0;
}


bool ShowRecommender::isEligible(const Show& s) const
{
    // DEBUG 3B: Output candidate evaluation details
    bool inList = inLists_.contains(s.traktId);
    qDebug() << "[DEBUG Eligibility] Checking Show:" << s.title 
             << "ID:" << s.traktId 
             << "Is in inLists_ set?" << inList;

    if (s.traktId <= 0) {
        qDebug() << "Reject" << s.title << "reason: invalid traktId";
        return false;
    }
    if (s.status.compare("ended", Qt::CaseInsensitive) != 0) {
        qDebug() << "Reject" << s.title << "reason: not ended, status=" << s.status;
        return false;
    }
    if (s.status.compare("canceled", Qt::CaseInsensitive) == 0) {
        qDebug() << "Reject" << s.title << "reason: canceled";
        return false;
    }
    if (inList) {
        qDebug() << "Reject" << s.title << "reason: in user lists";
        return false;
    }
    if (watched_.contains(s.traktId)) {
        qDebug() << "Reject" << s.title << "reason: watched";
        return false;
    }
    if (state_->isSkipped(s.traktId)) {
        qDebug() << "Reject" << s.title << "reason: manually skipped";
        return false;
    }
    // In ShowRecommender::isEligible
    if (s.language.compare("en", Qt::CaseInsensitive) != 0) {
        qDebug() << "Reject" << s.title << "reason: non-English language (" << s.language << ")";
        return false;
    }
    return true;
}

Show ShowRecommender::fallbackRelated()
{
    const QMap<QString, QString>& lists = client_->listMap();

    for (int attempt = 0; attempt < cycle_.size(); ++attempt) {
        QString key = cycle_.at(cycleIndex_);
        cycleIndex_ = (cycleIndex_ + 1) % cycle_.size();

        if (!lists.contains(key)) continue;

        QString slug = lists.value(key);
        QVector<Show> listShows = client_->fetchListShows(slug);
        if (listShows.isEmpty()) continue;

        int randomIndex = QRandomGenerator::global()->bounded(listShows.size());
        int seedId = listShows[randomIndex].traktId;

        QVector<Show> related = client_->fetchRelatedShows(seedId);

        for (Show& s : related) {
            if (isEligible(s)) {
                // Populate heavy details ONLY for the selected fallback show!
                client_->populateShowDetails(s);
                return s;
            }
        }
    }

    Show empty;
    empty.title = "No more recommendations";
    empty.overview = "All lists exhausted.";
    return empty;
}


Show ShowRecommender::nextShow()
{
    ensureLoaded();

    qDebug() << "Candidates loaded:" << candidates_.size();
    qDebug() << "inLists size:" << inLists_.size();
    qDebug() << "watched size:" << watched_.size();

    while (index_ < candidates_.size()) {
        Show s = candidates_.at(index_);
        index_++;

        bool ok = isEligible(s);
        qDebug() << "Checking candidate:" << s.traktId << s.title
                 << "status=" << s.status
                 << "eligible=" << ok;

        if (ok) {
            // Populate expensive details (sampled runtime & episodes) ONLY when selected!
            client_->populateShowDetails(s);
            return s;
        }
    }

    qDebug() << "Falling back to related";
    return fallbackRelated();
}

File: website/classes/ShowRecommender.php
<?php

require_once __DIR__ . '/TVMazeClient.php';
require_once __DIR__ . '/StateStore.php';

class ShowRecommender
{
    private TVMazeClient $client;
    private StateStore $state;

    private array $candidates = [];
    private int $index = 0;

    private array $inLists = [];
    private array $skipped = [];

    private array $cycle = ['30', '40', '60', 'sleepy', 'sitcom'];
    private int $cycleIndex = 0;

    public function __construct(TVMazeClient $client, StateStore $state)
    {
        $this->client = $client;
        $this->state = $state;
    }

    public function markAsInList(int $showId): void
    {
        if (!in_array($showId, $this->inLists, true)) {
            $this->inLists[] = $showId;
        }
    }

    public function markAsSkipped(int $showId): void
    {
        if (!in_array($showId, $this->skipped, true)) {
            $this->skipped[] = $showId;
        }
    }

    private function ensureLoaded(): void
    {
        $this->inLists = $this->state->getAllListShowIds();
        $this->skipped = $this->state->getSkippedShowIds();

        if (empty($this->candidates)) {
            $fetched = $this->client->fetchCandidateShows();
            shuffle($fetched);
            $this->candidates = $fetched;
            $this->index = 0;
        }
    }

    public function isEligible(array $show): bool
    {
        $id = (int)($show['id'] ?? 0);
        if ($id <= 0) {
            return false;
        }

        $status = strtolower(trim((string)($show['status'] ?? '')));
        if ($status !== 'ended') {
            return false;
        }

        $lang = strtolower(trim((string)($show['language'] ?? '')));
        if ($lang !== 'english' && $lang !== 'en') {
            return false;
        }

        if (in_array($id, $this->inLists, true)) {
            return false;
        }

        if (in_array($id, $this->skipped, true)) {
            return false;
        }

        return true;
    }

    private function findRelatedShow(): ?array
    {
        $cycleCount = count($this->cycle);
        $this->cycleIndex = $this->state->getCycleIndex() % $cycleCount;

        for ($attempt = 0; $attempt < $cycleCount; $attempt++) {
            $key = $this->cycle[$this->cycleIndex];
            $nextCycleIndex = ($this->cycleIndex + 1) % $cycleCount;

            $listShows = $this->state->getShowsInList($key);
            if (empty($listShows)) {
                $this->cycleIndex = $nextCycleIndex;
                continue;
            }

            // Shuffle list shows to vary seed show pick
            shuffle($listShows);
            foreach ($listShows as $seedShow) {
                $seedId = (int)($seedShow['id'] ?? 0);
                if ($seedId <= 0) continue;

                $relatedShows = $this->client->fetchRelatedShows($seedId, $seedShow);
                shuffle($relatedShows);

                foreach ($relatedShows as $s) {
                    if ($this->isEligible($s)) {
                        $this->client->populateShowDetails($s);
                        $this->cycleIndex = $nextCycleIndex;
                        $this->state->setCycleIndex($this->cycleIndex);
                        return $s;
                    }
                }
            }

            $this->cycleIndex = $nextCycleIndex;
        }

        $this->state->setCycleIndex($this->cycleIndex);
        return null;
    }

    public function nextShow(): array
    {
        $this->ensureLoaded();

        // 1. Try finding a show related to items in user's lists first
        $related = $this->findRelatedShow();
        if ($related !== null) {
            return $related;
        }

        // 2. Fall back to candidates pool
        $total = count($this->candidates);
        while ($this->index < $total) {
            $s = $this->candidates[$this->index];
            $this->index++;

            if ($this->isEligible($s)) {
                $this->client->populateShowDetails($s);
                return $s;
            }
        }

        // 3. Fallback if exhausted
        return [
            'id' => 0,
            'title' => 'No more recommendations',
            'overview' => 'All lists and candidates exhausted.',
            'status' => 'Ended',
            'language' => 'English',
            'runtime' => 0,
            'posterUrl' => '',
            'genres' => [],
            'firstAired' => '',
            'seasonCount' => 0,
            'totalEpisodes' => 0
        ];
    }
}
